src/modules/types.rs

```diff
@@ -14,14 +14,24 @@ pub enum Type {
     Generic
 }
 
+impl Type {
+    pub fn is_subset(&self, of: &Type) -> bool {
+        match (self, of) {
+            (Type::Array(l), Type::Array(r)) if !matches!(**l, Type::Generic) => **r == Type::Generic,
+            (l, Type::Failable(r)) if !matches!(l, Type::Failable(_)) => *l == **r,
```

Comment:
i feel like this needs to be refactored for readability. i can't make out what happens here at all

Reply:
Does this code look any better to you?

```
match (self, of) {
    (Type::Array(l), Type::Array(r)) => {
        !matches!(**l, Type::Generic) && matches!(**r, Type::Generic)
    }
    (l, Type::Failable(r)) => !matches!(l, Type::Failable(_)) && *l == **r,
    _ => false
}
```